subsys/net/lib/sockets/sockets.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_SOCKETS)
+#define SYS_LOG_DOMAIN "net/sock"
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <net/net_context.h>
+#include <net/net_pkt.h>
+#include <net/socket.h>
+#include "sockets_int.h"
+
+#define SET_ERRNO(x) \
+	{ int _err = x; if (_err < 0) { errno = -_err; return -1; } }
+
+#define sock_is_eof(ctx) ((ctx)->user_data != NULL)
+#define sock_set_eof(ctx) { (ctx)->user_data = INT_TO_POINTER(1); }
+
+static inline void _k_fifo_wait_non_empty(struct k_fifo *fifo, int32_t timeout)
+{
+	struct k_poll_event events[] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+					 K_POLL_MODE_NOTIFY_ONLY, fifo),
+	};
+
+	k_poll(events, ARRAY_SIZE(events), timeout);
+}
+
+static void zsock_flush_queue(struct net_context *ctx)
+{
+	bool is_listen = net_context_get_state(ctx) == NET_CONTEXT_LISTENING;
+	void *p;
+
+	/* recv_q and accept_q are shared via a union */
+	while ((p = k_fifo_get(&ctx->recv_q, K_NO_WAIT)) != NULL) {
+		if (is_listen) {
+			NET_DBG("discarding ctx %p", p);
+			net_context_put(p);
+		} else {
+			NET_DBG("discarding pkt %p", p);
+			net_pkt_unref(p);
+		}
+	}
+}
+
+int zsock_socket(int family, int type, int proto)
+{
+	struct net_context *ctx;
+
+	SET_ERRNO(net_context_get(family, type, proto, &ctx));
+	/* recv_q and accept_q are in union */
+	k_fifo_init(&ctx->recv_q);
+
+	/* TODO: Ensure non-negative */
+	return POINTER_TO_INT(ctx);
+}
+
+int zsock_close(int sock)
+{
+	struct net_context *ctx = INT_TO_POINTER(sock);
+
+	/* Reset callbacks to avoid any race conditions while
+	 * flushing queues.
+	 */
+	net_context_accept(ctx, NULL, K_NO_WAIT, NULL);
+	net_context_recv(ctx, NULL, K_NO_WAIT, NULL);
+
+	zsock_flush_queue(ctx);
+
+	SET_ERRNO(net_context_put(ctx));
+	return 0;
+}
+
+static void zsock_accepted_cb(struct net_context *new_ctx,
+			      struct sockaddr *addr, socklen_t addrlen,
+			      int status, void *user_data) {
+	struct net_context *parent = user_data;
+
+	NET_DBG("parent=%p, ctx=%p, st=%d", parent, new_ctx, status);
+
+	k_fifo_put(&parent->accept_q, new_ctx);
+}
+
+static void zsock_received_cb(struct net_context *ctx, struct net_pkt *pkt,
+			      int status, void *user_data) {
+	unsigned int header_len;
+
+	NET_DBG("ctx=%p, pkt=%p, st=%d, user_data=%p", ctx, pkt, status,
+		user_data);
+
+	/* if pkt == NULL, EOF */
+	if (pkt == NULL) {
```

Comment:
It would be better to have "if (!pkt) {" here, so that the code has unified look with rest of the IP stack.